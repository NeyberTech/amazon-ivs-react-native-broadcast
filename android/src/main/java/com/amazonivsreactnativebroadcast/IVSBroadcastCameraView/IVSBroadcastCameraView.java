package com.amazonivsreactnativebroadcast.IVSBroadcastCameraView;

import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.amazonaws.ivs.broadcast.BroadcastSession;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;

public class IVSBroadcastCameraView extends LinearLayout implements LifecycleEventListener {
  public static final String START_COMMAND_NAME = "START";
  public static final String STOP_COMMAND_NAME = "STOP";
  @Deprecated
  public static final String SWAP_CAMERA_COMMAND_NAME = "SWAP_CAMERA";

  public enum Events {
    ON_BROADCAST_STATE_CHANGED("onBroadcastStateChanged"),
    ON_BROADCAST_ERROR("onBroadcastError"),
    ON_BROADCAST_QUALITY_CHANGED("onBroadcastQualityChanged"),
    ON_BROADCAST_AUDIO_STATS("onBroadcastAudioStats"),
    ON_IS_BROADCAST_READY("onIsBroadcastReady"),
    ON_NETWORK_HEALTH_CHANGED("onNetworkHealthChanged"),
    ON_ERROR("onError");

    private String title;

    Events(String title) {
      this.title = title;
    }

    @Override
    public String toString() {
      return title;
    }
  }

  private String STREAM_KEY;
  private String RTMPS_URL;
  private IVSBroadcastSessionService ivsBroadcastSession;

  private void addCameraPreview(@NonNull View preview) {
    preview.layout(0, 0, getRight(), getBottom());
    addView(preview);
  }

  private void onReceiveCameraPreviewHandler(View preview) {
    removeAllViews();
    addCameraPreview(preview);
  }

  private void sendEvent(String eventName, @Nullable WritableMap eventPayload) {
    ThemedReactContext reactContext = (ThemedReactContext) super.getContext();
    RCTEventEmitter eventEmitter = reactContext.getJSModule(RCTEventEmitter.class);
    eventEmitter.receiveEvent(getId(), eventName, eventPayload);
  }

  private void sendErrorEvent(String errorMessage) {
    WritableMap eventPayload = Arguments.createMap();
    eventPayload.putString("message", errorMessage);

    sendEvent(Events.ON_ERROR.toString(), eventPayload);
  }

  private void initBroadcastSession() {
    if (ivsBroadcastSession.isInitialized()) return;

    try {
      IVSBroadcastSessionListener broadcastSessionListener = new IVSBroadcastSessionListener(
        this::sendEvent
      );
      ivsBroadcastSession.setListener(broadcastSessionListener.listener);

      String broadcastSessionId = ivsBroadcastSession.init();
      broadcastSessionListener.setBroadcastSessionId(broadcastSessionId);

      ivsBroadcastSession.getCameraPreviewAsync((cameraPreview) -> {
        onReceiveCameraPreviewHandler(cameraPreview);

        WritableMap eventPayload = Arguments.createMap();
        eventPayload.putBoolean("isReady", ivsBroadcastSession.isReady());
        sendEvent(Events.ON_IS_BROADCAST_READY.toString(), eventPayload);
      });
    } catch (RuntimeException error) {
      sendErrorEvent(error.toString());
    }
  }

  public IVSBroadcastCameraView(ThemedReactContext reactContext) {
    super(reactContext);
    reactContext.addLifecycleEventListener(this);
    ivsBroadcastSession = new IVSBroadcastSessionService(reactContext);
  }

  protected void start() {
    if (RTMPS_URL != null && STREAM_KEY != null) {
      try {
        ivsBroadcastSession.start(RTMPS_URL, STREAM_KEY);
      } catch (RuntimeException error) {
        sendErrorEvent(error.toString());
      }
    } else {
      throw new RuntimeException("'rtmpsUrl' and 'streamKey' props are required!");
    }
  }

  protected void stop() {
    try {
      ivsBroadcastSession.stop();
    } catch (RuntimeException error) {
      sendErrorEvent(error.toString());
    }
  }

  @Deprecated
  protected void swapCamera() {
    try {
      ivsBroadcastSession.swapCamera(this::onReceiveCameraPreviewHandler);
    } catch (RuntimeException error) {
      sendErrorEvent(error.toString());
    }
  }

  protected void cleanUp() {
    removeAllViews();
    ivsBroadcastSession.releaseResources();
  }

  protected void setIsMuted(boolean isMuted) {
    ivsBroadcastSession.setIsMuted(isMuted);
  }

  protected void setIsCameraPreviewMirrored(boolean isCameraPreviewMirrored) {
    ivsBroadcastSession.setIsCameraPreviewMirrored(isCameraPreviewMirrored, this::onReceiveCameraPreviewHandler);
  }

  protected void setCameraPosition(String cameraPosition) {
    ivsBroadcastSession.setCameraPosition(cameraPosition, this::onReceiveCameraPreviewHandler);
  }

  protected void setCameraPreviewAspectMode(String cameraPreviewAspectMode) {
    ivsBroadcastSession.setCameraPreviewAspectMode(cameraPreviewAspectMode, this::onReceiveCameraPreviewHandler);
  }

  protected void setRtmpsUrl(String rtmpsUrl) {
    RTMPS_URL = rtmpsUrl;
  }

  protected void setStreamKey(String streamKey) {
    STREAM_KEY = streamKey;
  }

  protected void setSessionLogLevel(String sessionLogLevel) {
    ivsBroadcastSession.setSessionLogLevel(sessionLogLevel);
  }

  protected void setLogLevel(String logLevel) {
    ivsBroadcastSession.setLogLevel(logLevel);
  }

  protected void setConfigurationPreset(String configurationPreset) {
    ivsBroadcastSession.setConfigurationPreset(configurationPreset);
  }

  protected void setVideoConfig(ReadableMap videoConfig) {
    ivsBroadcastSession.setVideoConfig(videoConfig);
  }

  protected void setAudioConfig(ReadableMap audioConfig) {
    ivsBroadcastSession.setAudioConfig(audioConfig);
  }

  @Override
  public void onHostResume() {
  }

  @Override
  public void onHostPause() {
  }

  @Override
  public void onHostDestroy() {
    cleanUp();
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    initBroadcastSession();
  }
}
