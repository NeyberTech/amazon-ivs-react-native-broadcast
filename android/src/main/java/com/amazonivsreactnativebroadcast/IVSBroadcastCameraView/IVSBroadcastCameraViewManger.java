package com.amazonivsreactnativebroadcast.IVSBroadcastCameraView;

import java.util.*;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import javax.annotation.Nullable;

public class IVSBroadcastCameraViewManger extends ViewGroupManager<IVSBroadcastCameraView> {
  @Override
  public String getName() {
    return "RCTIVSBroadcastCameraView";
  }

  @Override
  public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
    MapBuilder.Builder<String, Object> builder = MapBuilder.<String, Object>builder();
    for (IVSBroadcastCameraView.Events event : IVSBroadcastCameraView.Events.values()) {
      builder.put(event.toString(), MapBuilder.of("registrationName", event.toString()));
    }
    return builder.build();
  }

  @Nullable
  @Override
  public Map<String, Integer> getCommandsMap() {
    return MapBuilder.of(IVSBroadcastCameraView.START_COMMAND_NAME, 0,
      IVSBroadcastCameraView.STOP_COMMAND_NAME, 1,
      IVSBroadcastCameraView.SWAP_CAMERA_COMMAND_NAME, 2);
  }

  @Override
  public void receiveCommand(IVSBroadcastCameraView view, String commandId, @Nullable ReadableArray args) {
    switch (commandId) {
      case IVSBroadcastCameraView.START_COMMAND_NAME: {
        view.start();
        break;
      }
      case IVSBroadcastCameraView.STOP_COMMAND_NAME: {
        view.stop();
        break;
      }
      case IVSBroadcastCameraView.SWAP_CAMERA_COMMAND_NAME: {
        view.swapCamera();
        break;
      }
      default: {
        throw new RuntimeException("The following command is not supported yet: " + commandId);
      }
    }
  }

  @Override
  protected IVSBroadcastCameraView createViewInstance(ThemedReactContext reactContext) {
    return new IVSBroadcastCameraView(reactContext);
  }

  @Override
  public void onDropViewInstance(IVSBroadcastCameraView view) {
    super.onDropViewInstance(view);
    view.cleanUp();
  }

  @ReactProp(name = "isCameraPreviewMirrored")
  public void setIsCameraPreviewMirrored(IVSBroadcastCameraView view, boolean isCameraPreviewMirrored) {
    view.setIsCameraPreviewMirrored(isCameraPreviewMirrored);
  }

  @ReactProp(name = "cameraPosition")
  public void setCameraPosition(IVSBroadcastCameraView view, String cameraPosition) {
    view.setCameraPosition(cameraPosition);
  }

  @ReactProp(name = "cameraPreviewAspectMode")
  public void setCameraPreviewAspectMode(IVSBroadcastCameraView view, String cameraPreviewAspectMode) {
    view.setCameraPreviewAspectMode(cameraPreviewAspectMode);
  }

  @ReactProp(name = "rtmpsUrl")
  public void setRtmpsUrl(IVSBroadcastCameraView view, String rtmpsUrl) {
    view.setRtmpsUrl(rtmpsUrl);
  }

  @ReactProp(name = "streamKey")
  public void setStreamKey(IVSBroadcastCameraView view, String streamKey) {
    view.setStreamKey(streamKey);
  }

  @ReactProp(name = "logLevel")
  public void setLogLevel(IVSBroadcastCameraView view, String logLevel) {
    view.setLogLevel(logLevel);
  }

  @ReactProp(name = "sessionLogLevel")
  public void setSessionLogLevel(IVSBroadcastCameraView view, String sessionLogLevel) {
    view.setSessionLogLevel(sessionLogLevel);
  }

  @ReactProp(name = "videoConfig")
  public void setVideoConfig(IVSBroadcastCameraView view, ReadableMap videoConfig) {
    view.setVideoConfig(videoConfig);
  }

  @ReactProp(name = "audioConfig")
  public void setAudioConfig(IVSBroadcastCameraView view, ReadableMap audioConfig) {
    view.setAudioConfig(audioConfig);
  }
}
