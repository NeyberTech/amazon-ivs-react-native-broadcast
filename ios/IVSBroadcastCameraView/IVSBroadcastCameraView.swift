import UIKit
import AVFoundation

class IVSBroadcastCameraView: UIView {
  private var wasInitialSubViewAdded: Bool = false
  private let broadcastSession: IVSBroadcastSessionService = IVSBroadcastSessionService()
  
  @objc var streamKey: NSString?
  @objc var rtmpsUrl: NSString?
  @objc var isCameraPreviewMirrored: Bool = false
  @objc var cameraPreviewAspectMode: NSString = "none"
  @objc var sessionLogLevel: NSString?
  @objc var cameraPosition: NSString? {
    didSet {
      self.broadcastSession.setCameraPosition(cameraPosition)
    }
  }
  @objc var logLevel: NSString? {
    didSet {
      self.broadcastSession.setLogLevel(logLevel)
    }
  }
  @objc var videoConfig: NSDictionary? {
    didSet {
      do {
        try self.broadcastSession.setVideoConfig(videoConfig)
      } catch {
        self.onErrorHandler(error)
      }
    }
  }
  @objc var audioConfig: NSDictionary? {
    didSet {
      do {
        try self.broadcastSession.setAudioConfig(audioConfig)
      } catch {
        self.onErrorHandler(error)
      }
    }
  }
  
  @objc var onError: RCTDirectEventBlock?
  @objc var onIsBroadcastReady: RCTDirectEventBlock?
  @objc var onAudioSessionInterrupted: RCTDirectEventBlock?
  @objc var onAudioSessionResumed: RCTDirectEventBlock?
  @objc var onMediaServicesWereLost: RCTDirectEventBlock?
  @objc var onMediaServicesWereReset: RCTDirectEventBlock?
  @objc var onBroadcastError: RCTDirectEventBlock? {
    didSet {
      self.broadcastSession.setBroadcastErrorHandler(onBroadcastError)
    }
  }
  @objc var onBroadcastAudioStats: RCTDirectEventBlock? {
    didSet {
      self.broadcastSession.setBroadcastAudioStatsHandler(onBroadcastAudioStats)
    }
  }
  @objc var onBroadcastStateChanged: RCTDirectEventBlock? {
    didSet {
      self.broadcastSession.setBroadcastStateChangedHandler(onBroadcastStateChanged)
    }
  }
  @objc var onBroadcastQualityChanged: RCTDirectEventBlock? {
    didSet {
      self.broadcastSession.setBroadcastQualityChangedHandler(onBroadcastQualityChanged)
    }
  }
  @objc var onNetworkHealthChanged: RCTDirectEventBlock? {
    didSet {
      self.broadcastSession.setNetworkHealthChangedHandler(onNetworkHealthChanged)
    }
  }
  
  @objc
  private func audioSessionInterrupted(_ notification: Notification) {
    guard let userInfo = notification.userInfo,
          let typeValue = userInfo[AVAudioSessionInterruptionTypeKey] as? UInt,
          let type = AVAudioSession.InterruptionType(rawValue: typeValue)
    else { return }
    
    switch type {
      case .began:
        self.onAudioSessionInterrupted?([:])
      case .ended:
        self.onAudioSessionResumed?([:])
      @unknown default:
        break
    }
  }
  
  @objc
  private func mediaServicesWereLost(_ notification: Notification) {
    self.onMediaServicesWereLost?([:])
  }
  
  @objc
  private func mediaServicesWereReset(_ notification: Notification) {
    self.onMediaServicesWereReset?([:])
  }
  
  // Observing notifications sent through NSNotificationCenter (Audio Interruptions & Media Services Lost sections)
  private func subscribeToNotificationCenter() {
    let center = NotificationCenter.default
    center.addObserver(
      self,
      selector: #selector(audioSessionInterrupted(_:)),
      name: AVAudioSession.interruptionNotification,
      object: nil)
    center.addObserver(
      self,
      selector: #selector(mediaServicesWereLost(_:)),
      name: AVAudioSession.mediaServicesWereLostNotification,
      object: nil)
    center.addObserver(
      self,
      selector:  #selector(mediaServicesWereReset(_:)),
      name: AVAudioSession.mediaServicesWereResetNotification,
      object: nil)
  }
  
  private func unsubscribeNotificationCenter() {
    let center = NotificationCenter.default
    center.removeObserver(self, name: AVAudioSession.interruptionNotification, object: nil)
    center.removeObserver(self, name: AVAudioSession.mediaServicesWereLostNotification, object: nil)
    center.removeObserver(self, name: AVAudioSession.mediaServicesWereResetNotification, object: nil)
  }
  
  private func onErrorHandler(_ error: Error) {
    self.onError?(["message": error.localizedDescription])
  }
  
  private func onReceiveCameraPreviewHandler(error: Error?, preview: UIView?) {
    if error != nil {
      self.onErrorHandler(error!)
    } else if preview != nil  {
      self.subviews.forEach { $0.removeFromSuperview() }
      self.addSubview(preview!)
    } else {
      assertionFailure("Unexpected behaviour.")
    }
  }
  
  override init(frame: CGRect) {
    super.init(frame: frame)
  }
  
  required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented.")
  }
  
  override func didMoveToWindow() {
    if self.window != nil {
      // Disable the Application Idle Timer. Prevents device from going to sleep while using the broadcast SDK, which would interrupt the broadcast
      UIApplication.shared.isIdleTimerDisabled = true
      if self.broadcastSession.isInitiated() == false {
        self.subscribeToNotificationCenter()
        do {
          try self.broadcastSession.initiate()
          self.broadcastSession.setSessionLogLevel(self.sessionLogLevel)
          self.broadcastSession.getCameraPreviewAsync(aspectModeName: self.cameraPreviewAspectMode, isMirrored: self.isCameraPreviewMirrored, onReceiveCameraPreview: self.onReceiveCameraPreviewHandler)
        } catch {
          self.onErrorHandler(error)
        }
      }
    } else {
      UIApplication.shared.isIdleTimerDisabled = false
      unsubscribeNotificationCenter()
      self.subviews.forEach { $0.removeFromSuperview() }
      self.broadcastSession.deinitiate()
    }
  }
  
  override func didAddSubview(_ subview: UIView) {
    if self.wasInitialSubViewAdded == false {
      self.wasInitialSubViewAdded = true
      self.onIsBroadcastReady?(["isReady": self.broadcastSession.isReady()])
    }
  }
  
  public func start() {
    guard let rtmpsUrl = self.rtmpsUrl else {
      assertionFailure("'rtmpsUrl' prop is required.")
      return;
    }
    
    guard let streamKey = self.streamKey else {
      assertionFailure("'streamKey' prop is required.")
      return;
    }
    
    do {
      try self.broadcastSession.start(ivsRTMPSUrl: rtmpsUrl , ivsStreamKey: streamKey)
    } catch {
      self.onErrorHandler(error)
    }
  }
  
  public func stop() {
    self.broadcastSession.stop()
  }
  
  public func swapCamera() {
    do {
      try self.broadcastSession.swapCamera(aspectModeName: self.cameraPreviewAspectMode, isMirrored: self.isCameraPreviewMirrored, onReceiveCameraPreview: self.onReceiveCameraPreviewHandler)
    } catch {
      self.onErrorHandler(error)
    }
  }
}
