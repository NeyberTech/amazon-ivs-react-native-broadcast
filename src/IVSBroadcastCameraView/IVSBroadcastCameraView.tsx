import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import {
  Platform,
  UIManager,
  findNodeHandle,
  requireNativeComponent,
} from 'react-native';

import {
  Command,
  StateStatusEnum,
  StateStatusUnion,
  IIVSBroadcastCameraView,
  IIVSBroadcastCameraViewProps,
  IIVSBroadcastCameraNativeViewProps,
} from './IVSBroadcastCameraView.types';

const UNKNOWN = 'unknown';
export const NATIVE_VIEW_NAME = 'RCTIVSBroadcastCameraView';

const RCTIVSBroadcastCameraView =
  requireNativeComponent<IIVSBroadcastCameraNativeViewProps>(NATIVE_VIEW_NAME);

const NATIVE_SIDE_COMMANDS =
  UIManager.getViewManagerConfig(NATIVE_VIEW_NAME).Commands;

export const getCommandIdByPlatform = (command: Command) => {
  switch (Platform.OS) {
    case 'android': {
      return command;
    }
    case 'ios': {
      return NATIVE_SIDE_COMMANDS[command];
    }
    default: {
      return '';
    }
  }
};

const IVSBroadcastCameraView = forwardRef<
  IIVSBroadcastCameraView,
  IIVSBroadcastCameraViewProps
>((props, parentRef) => {
  const {
    onError,
    onBroadcastError,
    onIsBroadcastReady,
    onBroadcastAudioStats,
    onBroadcastStateChanged,
    onBroadcastQualityChanged,
    onNetworkHealthChanged,
    onAudioSessionInterrupted,
    onAudioSessionResumed,
    onMediaServicesWereLost,
    onMediaServicesWereReset,
    isMuted = false,
    isCameraPreviewMirrored = false,
    cameraPosition = 'back',
    cameraPreviewAspectMode = 'none',
    logLevel = 'error',
    sessionLogLevel = 'error',
    ...restProps
  } = props;

  const nativeViewRef = useRef(null);

  useImperativeHandle<IIVSBroadcastCameraView, IIVSBroadcastCameraView>(
    parentRef,
    () => {
      const reactTag = findNodeHandle(nativeViewRef.current);

      const dispatchViewManagerCommand = (
        command: Command,
        ...params: unknown[]
      ) =>
        UIManager.dispatchViewManagerCommand(
          reactTag,
          getCommandIdByPlatform(command),
          params ?? []
        );

      return {
        start: (
          options: Parameters<IIVSBroadcastCameraView['start']>[number] = {}
        ) => dispatchViewManagerCommand(Command.Start, options),
        stop: () => dispatchViewManagerCommand(Command.Stop),
        /**
         * @deprecated in favor of {@link cameraPosition}
         */
        swapCamera: () => dispatchViewManagerCommand(Command.SwapCamera),
      };
    },
    []
  );

  const onErrorHandler: IIVSBroadcastCameraNativeViewProps['onError'] = ({
    nativeEvent,
  }) => onError?.(nativeEvent.message);

  const onBroadcastErrorHandler: IIVSBroadcastCameraNativeViewProps['onBroadcastError'] =
    ({ nativeEvent }) => {
      const { code, type, detail, source, isFatal, sessionId } =
        nativeEvent.exception;

      onBroadcastError?.({
        code: String(code) ?? UNKNOWN,
        type: type ?? UNKNOWN,
        source: source ?? UNKNOWN,
        detail: detail ?? '',
        isFatal: !!isFatal,
        sessionId: sessionId ?? UNKNOWN,
      });
    };

  const onIsBroadcastReadyHandler: IIVSBroadcastCameraNativeViewProps['onIsBroadcastReady'] =
    ({ nativeEvent }) => onIsBroadcastReady?.(nativeEvent.isReady);

  const onBroadcastAudioStatsHandler: IIVSBroadcastCameraNativeViewProps['onBroadcastAudioStats'] =
    ({ nativeEvent }) => onBroadcastAudioStats?.(nativeEvent.audioStats);

  const onBroadcastStateChangedHandler: IIVSBroadcastCameraNativeViewProps['onBroadcastStateChanged'] =
    ({ nativeEvent }) => {
      const { stateStatus: incomingStateStatus, metadata } = nativeEvent;
      const outcomingStateStatus = (
        typeof incomingStateStatus === 'number'
          ? StateStatusEnum[incomingStateStatus]
          : incomingStateStatus
      ) as StateStatusUnion;
      onBroadcastStateChanged?.(outcomingStateStatus, metadata);
    };

  const onNetworkHealthChangedHandler: IIVSBroadcastCameraNativeViewProps['onNetworkHealthChanged'] =
    ({ nativeEvent }) => onNetworkHealthChanged?.(nativeEvent.networkHealth);

  const onBroadcastQualityChangedHandler: IIVSBroadcastCameraNativeViewProps['onBroadcastQualityChanged'] =
    ({ nativeEvent }) => onBroadcastQualityChanged?.(nativeEvent.quality);

  const onAudioSessionInterruptedHandler: IIVSBroadcastCameraNativeViewProps['onAudioSessionInterrupted'] =
    () => onAudioSessionInterrupted?.();

  const onAudioSessionResumedHandler: IIVSBroadcastCameraNativeViewProps['onAudioSessionResumed'] =
    () => onAudioSessionResumed?.();

  const onMediaServicesWereLostHandler: IIVSBroadcastCameraNativeViewProps['onMediaServicesWereLost'] =
    () => onMediaServicesWereLost?.();

  const onMediaServicesWereResetHandler: IIVSBroadcastCameraNativeViewProps['onMediaServicesWereReset'] =
    () => onMediaServicesWereReset?.();

  return (
    <RCTIVSBroadcastCameraView
      testID={NATIVE_VIEW_NAME}
      {...restProps}
      ref={nativeViewRef}
      isMuted={isMuted}
      logLevel={logLevel}
      sessionLogLevel={sessionLogLevel}
      cameraPreviewAspectMode={cameraPreviewAspectMode}
      isCameraPreviewMirrored={isCameraPreviewMirrored}
      cameraPosition={cameraPosition}
      onError={onErrorHandler}
      onBroadcastError={onBroadcastErrorHandler}
      onIsBroadcastReady={onIsBroadcastReadyHandler}
      onBroadcastAudioStats={onBroadcastAudioStatsHandler}
      onBroadcastStateChanged={onBroadcastStateChangedHandler}
      onBroadcastQualityChanged={onBroadcastQualityChangedHandler}
      onNetworkHealthChanged={onNetworkHealthChangedHandler}
      onAudioSessionInterrupted={onAudioSessionInterruptedHandler}
      onAudioSessionResumed={onAudioSessionResumedHandler}
      onMediaServicesWereLost={onMediaServicesWereLostHandler}
      onMediaServicesWereReset={onMediaServicesWereResetHandler}
    />
  );
});

export default IVSBroadcastCameraView;
