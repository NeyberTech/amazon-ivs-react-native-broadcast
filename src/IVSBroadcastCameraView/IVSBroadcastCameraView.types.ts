import type { Component, ComponentType } from 'react';
import type { NativeSyntheticEvent, ViewStyle, StyleProp } from 'react-native';

export type ExtractComponentProps<T> = T extends
  | ComponentType<infer P>
  | Component<infer P>
  ? P
  : never;

export enum Command {
  Start = 'START',
  Stop = 'STOP',
  /**
   * @deprecated in favor of {@link CameraPosition}
   */
  SwapCamera = 'SWAP_CAMERA',
}

export enum StateStatusEnum {
  INVALID = 0,
  DISCONNECTED = 1,
  CONNECTING = 2,
  CONNECTED = 3,
  ERROR = 4,
}

export type StateStatusUnion = keyof typeof StateStatusEnum;

export type LogLevel = 'debug' | 'error' | 'info' | 'warning';

export type CameraPosition = 'front' | 'back';

export type CameraPreviewAspectMode = 'fit' | 'fill' | 'none';

type AudioChannel = 1 | 2;

type AudioQuality = 'minimum' | 'low' | 'medium' | 'high' | 'maximum';

type KeyframeInterval = 1 | 2 | 3 | 4 | 5;

type AudioSessionStrategy = 'recordOnly' | 'playAndRecord' | 'noAction';

type ConfigurationPreset =
  | 'standardPortrait'
  | 'standardLandscape'
  | 'basicPortrait'
  | 'basicLandscape';

type AutomaticBitrateProfile = 'conservative' | 'fastIncrease';

interface IEventHandler<T extends Record<string, unknown>> {
  (event: NativeSyntheticEvent<T>): void;
}

export interface IBroadcastSessionError {
  readonly code: string;
  readonly type: string;
  readonly source: string;
  readonly detail: string;
  readonly isFatal: boolean;
  readonly sessionId: string;
}

export interface IAudioStats {
  readonly peak: number;
  readonly rms: number;
}

interface IVideoConfig {
  readonly width?: number;
  readonly height?: number;
  readonly bitrate?: number;
  readonly targetFrameRate?: number;
  readonly keyframeInterval?: KeyframeInterval;
  readonly isBFrames?: boolean;
  readonly isAutoBitrate?: boolean;
  readonly maxBitrate?: number;
  readonly minBitrate?: number;
  readonly autoBitrateProfile?: AutomaticBitrateProfile;
}

interface IAudioConfig {
  readonly bitrate?: number;
  readonly channels?: AudioChannel;
  readonly audioSessionStrategy?: AudioSessionStrategy;
  readonly quality?: AudioQuality;
}

interface IConnectedStateMetadata {
  sessionId: string;
}

export type StateChangedMetadata = IConnectedStateMetadata;

export interface INativeEventHandlers {
  onError: IEventHandler<Readonly<{ message: string }>>;
  onBroadcastError: IEventHandler<
    Readonly<{
      exception: {
        readonly code?: number;
        readonly type?: string;
        readonly source?: string;
        readonly detail?: string;
        readonly isFatal?: boolean;
        readonly sessionId?: string;
      };
    }>
  >;
  onIsBroadcastReady: IEventHandler<Readonly<{ isReady: boolean }>>;
  onBroadcastAudioStats: IEventHandler<Readonly<{ audioStats: IAudioStats }>>;
  onBroadcastStateChanged: IEventHandler<
    Readonly<{
      stateStatus: StateStatusUnion | number;
      metadata?: StateChangedMetadata;
    }>
  >;
  onBroadcastQualityChanged: IEventHandler<Readonly<{ quality: number }>>;
  onNetworkHealthChanged: IEventHandler<Readonly<{ networkHealth: number }>>;
  onAudioSessionInterrupted(): void;
  onAudioSessionResumed(): void;
  onMediaServicesWereReset(): void;
  onMediaServicesWereLost(): void;
}

export interface IIVSBroadcastCameraNativeViewProps
  extends IBaseProps,
    INativeEventHandlers {
  readonly style?: StyleProp<ViewStyle>;
  readonly testID?: string;
}

interface IBaseProps {
  readonly rtmpsUrl?: string;
  readonly streamKey?: string;
  readonly configurationPreset?: ConfigurationPreset;
  readonly videoConfig?: IVideoConfig;
  readonly audioConfig?: IAudioConfig;
  readonly logLevel?: LogLevel;
  readonly sessionLogLevel?: LogLevel;
  readonly cameraPreviewAspectMode?: CameraPreviewAspectMode;
  readonly isCameraPreviewMirrored?: boolean;
  readonly cameraPosition?: CameraPosition;
  readonly isMuted?: boolean;
}

export interface IEventHandlers {
  onError?(errorMessage: string): void;
  onBroadcastError?(error: IBroadcastSessionError): void;
  onIsBroadcastReady?(isReady: boolean): void;
  onBroadcastAudioStats?(audioStats: IAudioStats): void;
  onBroadcastStateChanged?(
    stateStatus: StateStatusUnion,
    metadata?: StateChangedMetadata
  ): void;
  onBroadcastQualityChanged?(quality: number): void;
  onNetworkHealthChanged?(networkHealth: number): void;
  onAudioSessionInterrupted?(): void;
  onAudioSessionResumed?(): void;
  onMediaServicesWereLost?(): void;
  onMediaServicesWereReset?(): void;
}

export interface IIVSBroadcastCameraViewProps
  extends IBaseProps,
    IEventHandlers {
  readonly style?: StyleProp<ViewStyle>;
  readonly testID?: string;
}

type StartMethodOptions = Pick<IBaseProps, 'rtmpsUrl' | 'streamKey'>;

export interface IIVSBroadcastCameraView {
  start(options?: StartMethodOptions): void;
  stop(): void;
  /**
   * @deprecated in favor of {@link CameraPosition}
   */
  swapCamera(): void;
}
