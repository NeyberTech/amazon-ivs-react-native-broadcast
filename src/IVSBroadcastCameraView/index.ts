export {
  CameraPosition,
  CameraPreviewAspectMode,
  StateStatusUnion,
  IAudioStats,
  IBroadcastSessionError,
  IIVSBroadcastCameraView,
} from './IVSBroadcastCameraView.types';
export { default as IVSBroadcastCameraView } from './IVSBroadcastCameraView';
