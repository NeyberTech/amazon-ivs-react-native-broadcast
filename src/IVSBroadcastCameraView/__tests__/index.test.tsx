import { Platform } from 'react-native';

import type { Command } from '../IVSBroadcastCameraView.types';
import { getCommandIdByPlatform } from '../IVSBroadcastCameraView';

const STOP_COMMAND = 'STOP';

jest.mock('react-native', () => {
  const ReactNative = jest.requireActual('react-native');

  ReactNative.UIManager.getViewManagerConfig = () => {
    return {
      Commands: {
        [STOP_COMMAND]: 1,
      },
    };
  };

  return ReactNative;
});

jest.mock('react-native/Libraries/Utilities/Platform', () => ({
  OS: 'ios',
}));

describe('getCommandIdByPlatform function works as expected', () => {
  it('iOS platform', () => {
    expect(getCommandIdByPlatform(STOP_COMMAND as Command)).toBe(1);
  });

  it('Android platform', () => {
    Platform.OS = 'android';
    expect(getCommandIdByPlatform(STOP_COMMAND as Command)).toBe(STOP_COMMAND);
  });
});
