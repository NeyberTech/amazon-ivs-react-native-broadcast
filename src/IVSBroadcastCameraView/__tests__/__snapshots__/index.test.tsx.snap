// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IVSBroadcastCameraView component Snapshot matched 1`] = `
<RCTIVSBroadcastCameraView
  cameraPosition="back"
  cameraPreviewAspectMode="none"
  isCameraPreviewMirrored={false}
  isMuted={false}
  logLevel="error"
  onAudioSessionInterrupted={[Function]}
  onAudioSessionResumed={[Function]}
  onBroadcastAudioStats={[Function]}
  onBroadcastError={[Function]}
  onBroadcastQualityChanged={[Function]}
  onBroadcastStateChanged={[Function]}
  onError={[Function]}
  onIsBroadcastReady={[Function]}
  onMediaServicesWereLost={[Function]}
  onMediaServicesWereReset={[Function]}
  onNetworkHealthChanged={[Function]}
  onTransmissionStatisticsChanged={[Function]}
  rtmpsUrl=""
  sessionLogLevel="error"
  streamKey=""
  testID="RCTIVSBroadcastCameraView"
/>
`;
